### PR TITLE
Introduce "DiffTest" pattern

### PR DESCRIPTION
**Pull request by pgonzal**
We have a number of unit tests that generate an output file, and then compare the file against its expected contents.  I am formalizing this approach by:

- Introducing a new "DiffTest" API in **node-core-library**.
- Promoting the JsonFile and PackageJsonLookup utilities into **node-core-library** (which will be our first API review and publishing for that package)
- Enabling the CopyStaticAssetsTask so that the expected output files can be stored in the "test" folder; this avoids the need for a separate "testInputs" folder (although it's still unclear how to handle api-extractor's *.ts files)

### CODE REVIEW

File: libraries/api-documenter/package.json [hunk 1]
```diff
@@ -15,12 +15,17 @@
   },
   "devDependencies": {
     "@microsoft/node-library-build": "~3.3.0",
+    "@microsoft/node-core-library": "~0.1.3",
+    "@types/chai": "3.4.34",
     "@types/colors": "~1.1.3",
     "@types/es6-collections": "0.5.29",
     "@types/fs-extra": "0.0.37",
     "@types/js-yaml": "~3.9.1",
```
Comment (qz2017) [inline libraries/api-documenter/package.json]: Why is this one using ~?

File: libraries/api-documenter/src/test/MarkdownPageRenderer.test.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+describe('MarkdownPageRenderer', () => {
+  it('renders markdown', (done: MochaDone) => {
+
+    const diffTest: DiffTest = new DiffTest();
+
+    const outputFolder: string = diffTest.getFolderPath(__dirname, 'MarkdownPageRenderer');
+
+    const renderer: MarkdownPageRenderer = new MarkdownPageRenderer(outputFolder);
+    const domPage: IDomPage = Domifier.createPage('Test page', 'test-id');
+    const outputFilename: string = renderer.writePage(domPage);
```
Comment (qz2017) [inline libraries/api-documenter/src/test/MarkdownPageRenderer.test.ts]: When do you clean up the temp output files? Ideally we should not really write to the disk and only compare in memory.

File: libraries/node-core-library/src/DiffTest.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  /**
+   * Returns a folder in the temp directory where the unit test should write its output files
+   * to be diffed.
+   * @param unitTestDirName - the "__dirname" variable, evaluated in the context of the unit test
+   * @param testModule - the name of the class being unit tested
+   * @returns A fully qualified path of the folder where the unit test should write its output
+   */
+  public getFolderPath(unitTestDirName: string, testModule: string): string {
```
Comment (pgonzal) [inline libraries/node-core-library/src/DiffTest.ts]: (feedback) Consider writing to a memory buffer and only flushing to disk when the diffs don't match

File: core-build/gulp-core-build/src/tasks/copyStaticAssets/CopyStaticAssetsTask.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ */
+export interface ICopyStaticAssetsTaskConfig {
+  includeExtensions?: string[];
+  excludeExtensions?: string[];
+  includeFiles?: string[];
+  excludeFiles?: string[];
+}
+
+/**
+ * Configures the @microsoft/gulp-core-build-webpack task with some smart defaults based on the package configuration.
```
Comment (pgonzal) [inline core-build/gulp-core-build/src/tasks/copyStaticAssets/CopyStaticAssetsTask.ts]: >@microsoft/gulp-core-build-webpack  [](start = 18, length = 35)

(feedback) fix this

File: libraries/node-core-library/src/DiffTest.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import { PackageJsonLookup } from './PackageJsonLookup';
+
+/**
+ * Implements a unit testing strategy that generates output files, and then
+ * compares them against the expected input.  If the files are different, then
+ * the test fails.
+ *
+ * @alpha
+ */
+export class DiffTest {
```
Comment (pgonzal) [inline libraries/node-core-library/src/DiffTest.ts]: >DiffTest [](start = 13, length = 8)

(feedback) make static and add flush cache function

File: libraries/node-core-library/src/DiffTest.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  /**
+   * Compares the contents of two files, and returns true if they are equivalent.
+   * Note that these files are not normally edited by a human; the "equivalence"
+   * comparison here is intended to ignore spurious changes that might be introduced
+   * by a tool, e.g. Git newline normalization or an editor that strips
+   * whitespace when saving.
+   */
+  public assertFileMatchesExpected(actualFilePath: string, expectedFilePath: string): void {
```
Comment (pgonzal) [inline libraries/node-core-library/src/DiffTest.ts]: >assertFileMatchesExpected [](start = 9, length = 25)

(feedback) simplify to 

fileDiffTest.assertEqual()

**Comment by pgonzal** [comment]
>   public tryFindPackagePathUpwards(sourceFilePath: string): string | undefined {

(feedback) tryGetPackageFolder

---
Refers to: libraries/node-core-library/src/PackageJsonLookup.ts:45 in 2791c31. [](commit_id = 2791c31dd185ff780ef2b2c2b16f71dd4dedc23b, deletion_comment = False)

**Comment by pgonzal** [comment]
>   public readPackageName(packageJsonPath: string): string {

(feedback) getPackageName()

---
Refers to: libraries/node-core-library/src/PackageJsonLookup.ts:74 in 2791c31. [](commit_id = 2791c31dd185ff780ef2b2c2b16f71dd4dedc23b, deletion_comment = False)

File: libraries/node-core-library/src/JsonFile.ts [hunk 7]
```diff
@@ -1,9 +1,107 @@
...
+        if (value === undefined) {
+          throw new Error(`The key "${key}" is undefined`);
+        }
+        JsonFile._validateNoUndefinedMembers(value);
+      }
+    }
+
+  }
+
+  private static _formatErrorDetails(errorDetails: Validator.SchemaErrorDetail[], indent: string,
```
Comment (pgonzal) [inline libraries/node-core-library/src/JsonFile.ts]: >_formatErrorDetails [](start = 17, length = 19)

(feedback) find the updated implementation that prints the rule name

Comment (pgonzal) [inline libraries/node-core-library/src/JsonFile.ts]: Get rid of all the other copies of Jsonfile

---
In reply to: [135151384](https://github.com/Microsoft/web-build-tools/pull/330#discussion_r135151384) [](ancestors = 135151384)

File: libraries/node-core-library/src/JsonFile.ts [hunk 8]
```diff
@@ -1,9 +1,107 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
+import * as fsx from 'fs-extra';
+import * as os from 'os';
+import * as jju from 'jju';
+import Validator = require('z-schema');
+
+export type ValidateErrorCallback = (errorDescription: string) => void;
+
```
Comment (pgonzal) [inline libraries/node-core-library/src/JsonFile.ts]: document

File: libraries/node-core-library/src/JsonFile.ts [hunk 9]
```diff
@@ -1,9 +1,107 @@
...
+    const buffer: Buffer = fsx.readFileSync(jsonFilename);
+    try {
+      return jju.parse(buffer.toString());
+    } catch (error) {
+      throw new Error(`Error reading "${jsonFilename}":` + os.EOL + `  ${error.message}`);
+    }
+  }
+
+  public static saveJsonFile(jsonFilename: string, jsonData: {}): void {
```
Comment (pgonzal) [inline libraries/node-core-library/src/JsonFile.ts]: >saveJsonFile [](start = 16, length = 12)

(feedback) add option for customizing newline

File: libraries/api-documenter/package.json (hunk 1 above)
Comment (pgonzal) [reply libraries/api-documenter/package.json]: Good catch

---
In reply to: [134927225](https://github.com/Microsoft/web-build-tools/pull/330#discussion_r134927225) [](ancestors = 134927225)

File: libraries/api-documenter/src/test/MarkdownPageRenderer.test.ts (hunk 2 above)
Comment (pgonzal) [reply libraries/api-documenter/src/test/MarkdownPageRenderer.test.ts]: About cleaning up the output files:  Good feedback.  I have renamed getFolderPath() --> prepareFolder() which always deletes the entire folder contents.


About in-memory comparisons, we discussed that at the API review.  It is a good idea, but a more complicated implementation:

- The caller must provide a buffer or write stream.
- The files do still need to get written to disk when a diff is encountered
- Sometimes a developer wants to see the output files anyway; maybe we need a switch for that.

Let's introduce it later.

---
In reply to: [134927460](https://github.com/Microsoft/web-build-tools/pull/330#discussion_r134927460) [](ancestors = 134927460)

File: libraries/node-core-library/src/JsonSchema.ts [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+   * NOTE: An error occurs if the file does not exist; however, the file itself is not loaded or validated
+   * until it the schema is actually used.
+   */
+  public static fromLoadedObject(schemaObject: Object): JsonSchema {
+    const schema: JsonSchema = new JsonSchema();
+    schema._schemaObject = schemaObject;
+    return schema;
+  }
+
+  private static _collectDependentSchemas(collectedSchemas: JsonSchema[], dependentSchemas: JsonSchema[],
```
Comment (qz2017) [inline libraries/node-core-library/src/JsonSchema.ts]: >_collectDependentSchemas [](start = 17, length = 24)

Document that diamond dependencies aren't supported

**Comment by pgonzal** [comment]
:shipit:

**Pull request merged**